tests/ui/use_self.rs

```diff
@@ -542,3 +542,42 @@ mod use_self_in_pat {
         }
     }
 }
+
+mod issue8845 {
+    pub enum Something {
+        Num(u8),
+        TupleNums(u8, u8),
+        StructNums { one: u8, two: u8 },
+    }
+
+    pub struct Foo(u8);
+
+    pub struct Bar {
+        x: u8,
+        y: usize,
+    }
+
+    impl Something {
+        const fn get_value(&self) -> u8 {
+            match self {
+                Something::Num(n) => *n,
+                Something::TupleNums(n, _m) => *n,
+                Something::StructNums { one, two: _ } => *one,
+            }
+        }
+    }
```

Comment:
Using the variants directly trigger the lint, suggesting to replace them all with `Self`:

```rust
fn imported_variants(&self) -> u8 {
    use Something::*;

    match self {
        Num(n) => *n,
        TupleNums(n, _m) => *n,
        StructNums { one, two: _ } => *one,
    }
}
```

Reply:
Seems that's uncovering an existing bug, the first path segment is picked unconditionally 

```rust
enum Enum {
    A,
}

impl Enum {
    fn f(self) -> usize {
        match self {
            crate::Enum::A => 1,
        }
    }
}
```

```
warning: unnecessary structure name repetition
  --> t.rs:13:13
   |
13 |             crate::Enum::A => 1,
   |             ^^^^^ help: use the applicable keyword: `Self`
   |
```

You could check the `res` field of the `PathSegment`s to pick the right one

Reply:
This was tricky. Using the `res` field on the `PathSegment`s doesn't seem to work. In the playground, I found that most of the `res` fields had errors, so I ended up using a workaround similar to what's done in the `check_expr` method. Luckily, the `lint_path_to_variant` function already exists. But, it's a bit hacky, and I did find that the name of the enum is correctly shown in the `res` fields. I made a version of the function which uses the `res` field, and is a bit less hacky, but it's also a bit less performant. I doubt anyone would notice the difference, but I left it unchanged for now.